### PR TITLE
runtime: start lifting contract preparation up through tx runtime layers 

### PR DESCRIPTION
**Pull request by nagisa**
Best reviewed commit-by-commit.

This ultimately lifts the contract preparation up through a several function call layers in the transaction runtime up to the layer where all the currently necessary data are available.

This PR also establishes in broad strokes where the pipelining decisions will be made (`ApplyProcessingReceiptState`) and makes some minor changes to the type to have it contain local receipts (in addition to the previously contained delayed receipts etc) in a queue of sorts which would allow the pipelining code to look-ahead of the ongoing processing work and queue-up preparation of the upcoming contracts there.

This work so far is intended to have no functional changes.

Part of #11319

### CODE REVIEW

**Comment by wacban** [comment]
@nagisa Hey, sorry, I won't have time to review this PR today, I'm going on PTO tomorrow and I'll be gone for a week. Can you find another reviewer or ping me again in a week?

**Comment by nagisa** [comment]
I'll find somebody else, enjoy your time off!

**Comment by frol** [comment]
@nagisa Sorry for jumping in, I'd like to invite you to participate in the [Race of Sloths](https://race-of-sloths.com). Just mention @race-of-sloths user in your github comment or PR description to join!

See how the flow works here: https://github.com/near/nearcore/pull/11778

**Comment by nagisa** [comment]
@race-of-sloths

**Comment by race-of-sloths** [comment]
@nagisa Thank you for your contribution! Your pull request is now a part of the [Race of Sloths](https://race-of-sloths.com)!
New Sloth joined the Race! Welcome!

[<picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://badge.race-of-sloths.com/nagisa?type=bot&theme=dark">
    <source media="(prefers-color-scheme: light)" srcset="https://badge.race-of-sloths.com/nagisa?type=bot&theme=light">
    Shows profile picture for the author of the PR
</picture>](https://race-of-sloths.com/profile/nagisa)


<details><summary>Current status: <i>waiting for scoring</i></summary>

We're waiting for maintainer to score this pull request with `@race-of-sloths score [0,1,2,3,5,8,13]` command
</details><details><summary>What is the Race of Sloths</summary>
<p>
Race of Sloths is a friendly competition where you can participate in challenges and compete with other open-source contributors within your normal workflow

For contributors:

- Tag @race-of-sloths inside your pull requests
- Wait for the maintainer to review and score your pull request
- Check out your position in the [Leaderboard](https://race-of-sloths.com/leaderboard)
- Keep weekly and monthly streaks to reach higher positions
- Boast your contributions with a dynamic picture of your [Profile](https://race-of-sloths.com/profile/nagisa)

For maintainers:
- Score pull requests that participate in the Race of Sloths
- Engage contributors with fair scoring and fast responses so they keep their streaks
- Promote the Race to the point where the Race starts promoting you
- Grow the community of your contributors

Feel free to check [our website](https://race-of-sloths.com) for additional details!
</p>
</details> 

<details>
<summary>Bot commands</summary>

- For contributors
  - **Include a PR:** `@race-of-sloths include` to enter the Race with your PR
- For maintainers:
  - **Assign points:** `@race-of-sloths score [1/2/3/5/8/13]` to award points based on your assessment.
  - **Reject this PR:** `@race-of-sloths exclude` to send this PR back to the drawing board.
  - **Exclude repo:** `@race-of-sloths pause` to stop bot activity in this repo until `@race-of-sloths unpause` command is called

</details>

File: runtime/runtime/src/ext.rs
Comment (tayfunelmas) [inline runtime/runtime/src/ext.rs:364]: nit. assuming `Ext` suffix represents "external", what does external contract mean here? (I miss the context for the structures in `ext.rs`.

**Review by tayfunelmas — APPROVED** [review]
LGTM. I did not attempted to connect the commits in a bigger picture, but the changes in each commit makes sense.

File: runtime/runtime/src/ext.rs
Comment (nagisa) [reply runtime/runtime/src/ext.rs:364]: That's a good question. "External" here stands for the "External Interface to the Contract Runtime". For this structure it could really just also read "RuntimeContractGetter" or "ContractRuntimeReader" or "Accessor" or something similar to evoke the purpose of this data structure more clearly, but since we already have `RuntimeExt` as a precedent, I opted to go with a similar name here as well.

**Comment by race-of-sloths** [comment]
### 🔄 The PR has been merged.
>[!IMPORTANT]
This pull request is a part of the Race of Sloths and has not been scored yet. Scoring will close in 24 hours! 🕰️

**Comment by race-of-sloths** [comment]
### ✅ PR is finalized! 
Your contribution is much appreciated with a final score of 0!
You have received 10 (0 base + 10 weekly bonus) Sloth points for this contribution


Another weekly streak completed, well done @nagisa! To keep your weekly streak and get another bonus make pull request next week! Looking forward to see you in race-of-sloths

**Pull request merged**
